keps/sig-auth/3257-cluster-trust-bundles/README.md

```diff
@@ -563,6 +579,34 @@ For example, if I maintain `example.com/my-signer`, I can use the following stra
   canary object first, and assess the health of the canary workloads.
 * Once I am satisfied that the change is safe, I edit the live object.
 
+### Publishing the kube-apiserver Trust Bundle
+
+Today, the trust bundle that allows verifying kube-apiserver serving certificate(s)
+at its internal endpoints is distributed into every namespace using a configMap.
+This is so that it can be mounted along with the ServiceAccount token in order
+for the workloads to be able to communicate with the kube-apiserver.
+
+To avoid the need for these configMaps in the future, a ClusterTrustBundle API
+object will be created for all clusters:
+
+```yaml
+apiVersion: v1beta1
+kind: ClusterTrustBundle
+metadata:
+  name: kubernetes.io:kube-apiserver-server:live # TODO: should we use "active" instead of "live", maybe?
+  labels:
+    kubernetes.io/cluster-trust-bundle-version: live
+spec:
+  signerName: kubernetes.io/kube-apiserver-server
+  trustBundle: "<... PEM CA ...>"
+```
+
+This object is managed by the Kubernetes Controller Manager. It serves the same
+purpose as the `kube-root-ca.crt` configMap - to verify internal kube-apiserver
+endpoints. There is currently no in-tree signer designated for these purposes,
+and so the signer with name `kubernetes.io/kube-apiserver-server` is introduced
+along with this bundle.
```

Comment:
A good exercise to make sure we fully specify the meaning / behavior of this signer is to add the documentation here for this new signer that we would add to https://kubernetes.io/docs/reference/access-authn-authz/certificate-signing-requests/#kubernetes-signers.

Because this is encompassing existing serving certificates issued by other existing signers, we have less control over the exact shape, but we can say some things like it must be valid for in order to work, e.g.:

* Permitted x509 extensions - honors key usage and DNSName/IPAddress subjectAltName extensions. At least one DNS or IP subjectAltName must be present.
   * Conformance test requires resolution of `kubernetes.default.svc.$cluster-domain` - I'm not sure the current status of the ability to change the cluster domain from `cluster.local` ... maybe get network folks feedback on whether it makes sense to spec that the kube-apiserver certificate should (must?) contain the DNS subjectAltName `kubernetes.default.svc` and/or `kubernetes.default.svc.$cluster-domain`
* Required key usages (other key usages may be present) - ["key encipherment", "digital signature", "server auth"] or ["digital signature", "server auth"].


This line in the signerName docs would also need to change if we don't intend to have k-c-m sign these certs
> The kube-controller-manager implements [control plane signing](https://kubernetes.io/docs/reference/access-authn-authz/certificate-signing-requests/#signer-control-plane) for each of the built in signers.

Reply:
> Because this is encompassing existing serving certificates issued by other existing signers

It's encompassing certificates you can use to _verify kube-apiserver serving certificates_, but it's not restricted to bundling the serving certs, right?

If I were to maintain a cluster, I wouldn't actually want to put a serving certificate in this bundle at all. I don't want to be updating it every couple of weeks when my serving certs expire. Instead, I want to bundle signers of those serving certs so that I can rotate my KAS serving certs freely, without having to worry about what's going to happen to the KAS trust in all the pods in the cluster.

I don't think we can easily restrict the certs in the bundle, except for we could probably ban the use of the "client auth" EKU.

Note that your proposal would likely make the bundle unusable in OpenShift.

Reply:
I added a commit with a section that talks of validation of this bundle, please see if it makes sense to you. I haven't run the DNS/IP SAN part through the sig-networking folk yet.

Reply:
> It's encompassing certificates you can use to _verify kube-apiserver serving certificates_, but it's not restricted to bundling the serving certs, right?

I think Jordan is saying this new signer takes ownership of the existing API server serving certs.  So for example, we cannot say something like "the SAN list but be exactly kubernetes.default.svc and kubernetes.default.svc.$cluster-domain" because existing working clusters could have extra SANs in their serving certs.  I understand Jordan's comment to only be about the signer, and not the CTB contents.

Reply:
yes, this was clarified in a later comment - https://github.com/kubernetes/enhancements/pull/4791#discussion_r1734725162